crates/ra_assists/src/assists/move_bounds.rs

```diff
@@ -36,23 +35,23 @@ pub(crate) fn move_bounds_to_where_clause(mut ctx: AssistCtx<impl HirDatabase>)
         AssistId("move_bounds_to_where_clause"),
         "move_bounds_to_where_clause",
         |edit| {
-            let type_params = type_param_list.type_params().collect::<Vec<_>>();
-
-            for param in &type_params {
-                if let Some(bounds) = param.type_bound_list() {
-                    let colon = param
-                        .syntax()
-                        .children_with_tokens()
-                        .find(|it| it.kind() == COLON)
-                        .unwrap();
-                    let start = colon.text_range().start();
-                    let end = bounds.syntax().text_range().end();
-                    edit.delete(TextRange::from_to(start, end));
-                }
-            }
-
-            let predicates = type_params.iter().filter_map(build_predicate);
-            let where_clause = Make::<ast::WhereClause>::from_predicates(predicates);
+            let new_params = type_param_list
+                .type_params()
+                .filter(|it| it.type_bound_list().is_some())
+                .map(|type_param| {
+                    let without_bounds =
+                        AstEditor::new(type_param.clone()).remove_bounds().ast().clone();
+                    (type_param, without_bounds)
+                });
```

Comment:
We are not fully using `EditorApi` here, as dynamically-typed `parent` still makes this impossible. However, at least the removal of bounds looks nice now.

Reply:
I see, but I'm a little bit confused about `replace_descendants` and `into_text_edit`. It looks like we do the job twice.

Reply:
Yeah, that's true: we first apply the edits to the syntax node to get the new node, and than we *recover* the edits by diffing two trees. That does sound like that anecdote about mathematician who, to boil an already full kettle, first empties the kettle into the sink, to get a problem they already know how to solve. 

The benefit here is that we separate syntax tree manipulation from text manipulation, and that should increase the reusability. For example, `remove_bounds` is a nice self-contained chunk of code now, which doesn't depend on the assist context.